src/gt4py/next/iterator/type_inference.py

```diff
@@ -380,20 +385,54 @@ def _infer_shift_location_types(shift_args, offset_provider, constraints):
     return current_loc_in, current_loc_out
 
 
-class _TypeInferrer(eve.NodeTranslator):
+@dataclasses.dataclass
+class _TypeInferrer(eve.traits.VisitorWithSymbolTableTrait, eve.NodeTranslator):
     """Visit the full iterator IR tree, convert nodes to respective types and generate constraints."""
 
-    def __init__(self, offset_provider):
-        self.offset_provider = offset_provider
+    offset_provider: Optional[dict[str, Connectivity | Dimension]]
+    collected_types: dict[int, Type] = dataclasses.field(default_factory=dict)
+    constraints: set[tuple[Type, Type]] = dataclasses.field(default_factory=set)
 
-    def visit_SymRef(
-        self, node: ir.SymRef, constraints: set[tuple[Type, Type]], symtypes: dict[str, Type]
-    ) -> Type:
-        if node.id in symtypes:
-            res = symtypes[node.id]
-            if isinstance(res, LetPolymorphic):
-                return freshen(res.dtype)
-            return res
+    @classmethod
+    def apply(
+        cls,
+        node: ir.Node,
+        offset_provider: Optional[dict[str, Connectivity | Dimension]] = None,
+        reindex: bool = True,
+    ) -> dict[int, Type]:
+        """
+        Infer the types of the child expressions of a given iterator IR expression.
+
+        The result is a dictionary mapping the id of child nodes to their type.
+        """
+        # Collect preliminary types of all nodes and constraints on them
+        inferrer = cls(offset_provider=offset_provider)
+        inferrer.visit(node)
+
+        # Ensure dict order is pre-order of the tree
+        collected_types = dict(reversed(inferrer.collected_types.items()))
+
+        # Compute the most general type that satisfies all constraints
+        unified_types = unify(list(collected_types.values()), inferrer.constraints)
+
+        if reindex:
+            unified_types = reindex_vars(list(unified_types))
+
+        return {
+            id_: unified_type for id_, unified_type in zip(collected_types.keys(), unified_types)
+        }
```

Comment:
Just a question: it looks to me like this part doesn't really belong to the Visitor since it's just postprocessing the result of the visit using the type unifier utilities from a different module. Wouldn't it be better to move this postprocessing part to the public entry point, the `infer()` function (and maybe also rename the pass)?

Reply:
I thought about this while adding the apply here, but decided against it as the visitor itself uses the apply method (for let polymorphic functions in `visit_FunctionDefinition`). My perspective on the `TypeInferrer` is that its task includes unification of constraints. Does that make sense?

Reply:
I also saw that the visitor calls its own `apply()` but still think it makes more sense to strictly separate the visitor logic from the postprocessing, and call `infer()` when needed from the visitor. I agree that the type inference process includes the unification of constraints, but the `_TypeInferrer` visitor (which is not even public interface of the module) is just the implementation of one of the required steps, so implementation-wise makes sense to separate cleanly the different steps.

Reply:
Just calling `infer` from the visitor is not enough as we need not only the type of the node, but also the mapping id => node type for all child nodes. I put it into  `_infer_all` (which is what I previously had). One thing I noticed is that now the postprocessing can not be customized in subclasses of `_TypeInferrer` anymore. In any case this will likely change soon anyway when we make this public interface.